waltz-web/src/main/java/com/khartec/waltz/web/endpoints/api/InvolvementKindEndpoint.java

```diff
@@ -1,32 +1,110 @@
     private static final String BASE_URL = WebUtilities.mkPath("api", "involvement-kind");
 
     private final InvolvementKindService service;
+    private UserRoleService userRoleService;
 
     @Autowired
-    public InvolvementKindEndpoint(InvolvementKindService service) {
+    public InvolvementKindEndpoint(InvolvementKindService service, UserRoleService userRoleService) {
         checkNotNull(service, "service must not be null");
+        checkNotNull(userRoleService, "userRoleService cannot be null");
 
         this.service = service;
+        this.userRoleService = userRoleService;
     }
 
-
     @Override
     public void register() {
-        EndpointUtilities.getForList(BASE_URL, (request, response) -> service.findAll());
+
+        getForList(BASE_URL, (request, response) -> service.findAll());
+        getForDatum(mkPath(BASE_URL, ":id"), this::getByIdRoute );
+
+        // create
+        postForDatum(BASE_URL, this::createInvolvementKindRoute );
+
+        // update
+        putForDatum(BASE_URL, this::updateInvolvementKindRoute );
+
+        // delete
+        deleteForDatum(mkPath(BASE_URL, ":id"), this::deleteInvolvementKindRoute);
+
+    }
+
+
+    private InvolvementKind getByIdRoute(Request request, Response response) {
+        long id = getId(request);
+        return service.getById(id);
+    }
+
+
+    private InvolvementKind createInvolvementKindRoute(Request request, Response response) throws IOException {
+        ensureUserHasAdminRights(request);
+
+        InvolvementKind involvementKind = extractAndPopulateLastUpdatedData(request);
+
+        LOG.info("User: {} creating Involvement Kind: {}", involvementKind.lastUpdatedBy(), involvementKind);
+        return service.create(involvementKind);
+    }
+
+
+    private InvolvementKind updateInvolvementKindRoute(Request request, Response response) throws IOException {
+        ensureUserHasAdminRights(request);
+
+        InvolvementKind involvementKind = extractAndPopulateLastUpdatedData(request);
+
+        LOG.info("User: {} updating Involvement Kind: {}", involvementKind.lastUpdatedBy(), involvementKind);
+        return service.update(involvementKind);
+    }
+
+
+    private boolean deleteInvolvementKindRoute(Request request, Response response) {
+        ensureUserHasAdminRights(request);
+
+        long id = getId(request);
+        String username = getUsername(request);
+
+        LOG.info("User: {} removing Involvement Kind: {}", username, id);
+
+        return service.delete(id);
+    }
+
+
+    private void ensureUserHasAdminRights(Request request) {
+        requireRole(userRoleService, request, Role.ADMIN);
+    }
+
+
+    private InvolvementKind extractAndPopulateLastUpdatedData(Request request) throws IOException {
+        return ImmutableInvolvementKind
+                .copyOf(readBody(request, InvolvementKind.class))
+                .withLastUpdatedBy(getUsername(request))
+                .withLastUpdatedAt(DateTimeUtilities.nowUtc());
```

Comment:
Going to merge anyway - but perhaps this belongs in service layer ?

`svc.save(ik, username);`